File: app/src/main/java/com/example/digitalmonk/data/local/db/AppDatabase.java
package com.example.digitalmonk.data.local.db;

import android.content.Context;
import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;

import com.example.digitalmonk.data.local.db.dao.UsageLogDao;
import com.example.digitalmonk.service.vpn.heartbeat.VpnHeartBeatDao;
import com.example.digitalmonk.data.local.db.entity.UsageLogEntity;
import com.example.digitalmonk.service.vpn.heartbeat.VpnHeartBeatEntity;

/**
 * The sole Database class for the application.
 * All entities (UsageLogs, Heartbeats, etc.) are registered here.
 */
@Database(entities = {UsageLogEntity.class, VpnHeartBeatEntity.class}, version = 1, exportSchema = false)
public abstract class AppDatabase extends RoomDatabase {

    private static volatile AppDatabase INSTANCE;

    public abstract UsageLogDao usageLogDao();
    public abstract VpnHeartBeatDao vpnHeartBeatDao();

    /**
     * Standard Singleton pattern to provide access to the database.
     */
    public static AppDatabase getDatabase(final Context context) {
        if (INSTANCE == null) {
            synchronized (AppDatabase.class) {
                if (INSTANCE == null) {
                    INSTANCE = Room.databaseBuilder(context.getApplicationContext(),
                                    AppDatabase.class, "app_database") // Generic name for the SQLite file
                            .fallbackToDestructiveMigration()
                            .build();
                }
            }
        }
        return INSTANCE;
    }
}

File: app/src/main/java/com/curbme/app/data/local/db/AppDatabase.kt
package com.curbme.app.data.local.db

import android.content.Context
import androidx.room.Database
import androidx.room.Room.databaseBuilder
import androidx.room.RoomDatabase
import com.curbme.app.data.local.db.dao.AppBlockDao
import com.curbme.app.data.local.db.dao.AppGroupDao
import com.curbme.app.data.local.db.dao.AppUsageDao
import com.curbme.app.data.local.db.dao.ReelStatsDao
import com.curbme.app.data.local.db.dao.ReelUsageStatsDao
import com.curbme.app.data.local.db.dao.UsageLogDao
import com.curbme.app.data.local.db.dao.WebsiteStatsDao
import com.curbme.app.data.local.db.entity.AppBlockRule
import com.curbme.app.data.local.db.entity.AppGroupEntity
import com.curbme.app.data.local.db.entity.AppUsageEntity
import com.curbme.app.data.local.db.entity.ReelStatsEntity
import com.curbme.app.data.local.db.entity.ReelUsageStatsEntity
import com.curbme.app.data.local.db.entity.UsageLogEntity
import com.curbme.app.data.local.db.entity.WebsiteStatsEntity
import com.curbme.app.service.vpn.heartbeat.VpnHeartBeatDao
import com.curbme.app.service.vpn.heartbeat.VpnHeartBeatEntity
import kotlin.concurrent.Volatile

/**
 * The sole Database class for the application.
 * All entities (UsageLogs, Heartbeats, ReelStats, WebsiteStats, AppGroups, etc.) are registered here.
 */
@Database(
    entities = [
        UsageLogEntity::class,
        VpnHeartBeatEntity::class,
        AppBlockRule::class,
        AppUsageEntity::class,
        ReelStatsEntity::class,
        ReelUsageStatsEntity::class,
        WebsiteStatsEntity::class,
        AppGroupEntity::class
    ],
    version = 7,
    exportSchema = false,
)
abstract class AppDatabase : RoomDatabase() {
    abstract fun usageLogDao(): UsageLogDao
    abstract fun appUsageDao(): AppUsageDao
    abstract fun vpnHeartBeatDao(): VpnHeartBeatDao
    abstract fun appBlockDao(): AppBlockDao
    abstract fun reelStatsDao(): ReelStatsDao
    abstract fun reelUsageStatsDao(): ReelUsageStatsDao
    abstract fun websiteStatsDao(): WebsiteStatsDao
    abstract fun appGroupDao(): AppGroupDao

    companion object {
        @Volatile
        private var INSTANCE: AppDatabase? = null

        /**
         * Standard Singleton pattern to provide access to the database.
         */
        fun getDatabase(context: Context): AppDatabase {
            return INSTANCE ?: synchronized(AppDatabase::class.java) {
                INSTANCE ?: databaseBuilder(
                    context.applicationContext,
                    AppDatabase::class.java,
                    "app_database",
                )
                    .enableMultiInstanceInvalidation()
                    .fallbackToDestructiveMigrationOnDowngrade(dropAllTables = false)
                    .fallbackToDestructiveMigration(dropAllTables = true)
                    .build().also { INSTANCE = it }
            }
        }

        /**
         * Emergency utility to wipe the database files from disk.
         * Used to fix "Integrity Hash" crashes without uninstallation.
         */
        fun deleteDatabaseFile(context: Context) {
            INSTANCE?.close()
            INSTANCE = null
            context.deleteDatabase("app_database")
        }
    }
}
